### ISSUE TITLE
Support for 2FA

### ISSUE THREAD

**Issue by CHTJonas** [camdram/camdram#434]
Placeholder to discuss implementation of two-factor authentication. Less important considering we are now depreciating passwords and moving to external OAuth providers, but still perhaps important bearing in mind that people leave laptops unattended in the clubroom with Raven login details saved in their browser's credentials manager.

My thinking would be a simple TOTP implementation where users scan a QR code into an authenticator app on their mobile devices. More details [here](https://github.com/google/google-authenticator/wiki/Key-Uri-Format).

**Comment by philosophicles** [comment]
This is possibly just me being thick, but I don't see how 2FA helps in practice with the "laptop in the clubroom" scenario...? 

IMHO, leaving laptops unattended in a public space, especially without an OS password layer (screen locked) is on the user, not on us. It's great that the Clubroom (still) has sufficient trust/security that people aren't worried about the laptop physically being stolen, but it is up to users to think of digital security too. I don't think we need to go too far to take responsibility for that.

From a practical perspective - are you suggesting we get the confirmation back from the external OAuth provider that the user is who they say they are (potentially having completed a 2FA challenge in that provider's flow), and then we offer a 2FA challenge of our own on top? 

To tackle the "laptop in the clubroom" risk, I think we'd have to be doing the 2FA challenge on a very regular basis - whenever there's been some inactivity on the site? So not including any option to "not ask on this device again" (ever / for 30 days / 1 week) - which is an option most 2FA flows offer. That feels like it'd be too burdensome for most users.

**Comment by hoyes** [comment]
As a bit of background on the rough intention with authentication at the moment:
 * Once you're logged in on a machine, Camdram permanently knows who you are (for 1 year), for the purposes of displaying read-only personalised content (albeit there's not much of that at the moment). This is all cookie-based after the initial login flow, so still applies if say, you originally logged in with Raven but your Raven account was deactivated in the meantime.
 * If you want to do anything sensitive or "admin-y" (generally speaking, something that ends up in making database changes) then we request a "re-authentication", which forces the user to validate who they are with Raven, Google, Facebook or Twitter (or a password re-entry, for now).
 * A re-authenticated session should last 30 minutes. So if the laptop's been idle for more than 30 minutes then there shouldn't be much anyone can do.
 * A re-authentication event also resets the "1 year" countdown above.

But... the move to OAuth means we're now reliant on however the OAuth provider handles the re-authentication. From memory:
 * For Facebook, I set a flag that forces the user to reenter their Facebook password (whether or not they were previously logged in).
 * Something similar is possible with Raven, but not implemented
 * No idea whether such an option exists on Google or Twitter.

Like @philosophicles, I'm on the fence over how much this matters though. If someone is able to use Google and/or Twitter to reauthenticate with Camdram on a random laptop, then they can also view their Gmail and post tweets as they're both less strict about timing out sessions than we are at the moment. There are however some inconsistencies in the above that should maybe be addressed, and a lot of the above are Symfony defaults rather than being explicit choices persay.

Having said all that, no reason not to still add 2FA as another (re)authentication method if useful. https://github.com/jonasarts/google-authenticator-bundle/ might be an option for Symfony integration.

**Comment by philosophicles** [comment]
If you're suggesting adding 2FA as an _alternative_ to reauthenticating via FB/Google/Twitter/Raven - I actually like that idea more. It's not especially an extra security layer at that point, just an alternative - if people find it more convenient. Though it's also a more-secure alternative for some attack vectors, I realize. 

If we were going to view it that way, as an alternative for re-authentication rather than as well as re-authentication: I'm gonna float the idea of a push notification to smartphone rather than Google Authenticator TOTP approach. I completely realize that would be more work and expect this suggestion to be shot down, but: it's just way cooler and more convenient for most users. 

Google does this themselves now, of course, for logging into Google accounts if you have an Android phone on the same account; I think Apple do the same for iPhones. My work uses a paid authentication/SSO service called Okta which also uses push notifications heavily for re-authentication and that works great. And I think other things are switching to using it gradually. 

The huge downside is certainly that it needs a particular app to be installed on the device, and in our case, that's probably an app that doesn't exist at all yet (unless there are any generic services we can piggy-back on?).

**Comment by CHTJonas** [comment]
Yeah 2FA for reauthentication or as a further stage in a general login sounds good to me. In terms of 2FA apps that I know of, you've got [Google Authenticator](https://en.wikipedia.org/wiki/Google_Authenticator), [Authy](https://authy.com/features/) and there's probably more.

Whilst we're discussing this we could also look at implementing FIDO2/U2F challenge response at the same time (this would work well with hardware tokens like YubiKeys if anyone here as seen or used those) but that might be a step too far.
